jax/core.py

```diff
@@ -1680,10 +1678,47 @@ def diff_shape(s1: Shape, s2: Shape) -> Shape:
   return tuple(map(diff_dim, s1, s2))
 
 def divide_shape_sizes(s1: Shape, s2: Shape) -> DimSize:
-  """Returns an integer "i" s.t., i * size(s2) == size(s1).
-  Raises if there is no such integer."""
+  """Returns a dimension "d" s.t., size(s1) = d * size(s2).
+  Raises if there is no such dimension.
+  """
   s1 = s1 or (1,)
   s2 = s2 or (1,)
+  # Special case when the non-constant dimensions in s2 are a subset of those in s1
+  def split_constants(s1: Shape) -> Tuple[int, Sequence[DimSize]]:
+    """Split a shape into the product of constant dimensions and list of non constant ones."""
+    prod_constant = 1
+    non_constant = []
+    for d in s1:
+      if is_constant_dim(d):
+        prod_constant *= operator.index(d)
+      else:
+        non_constant.append(d)
+    return prod_constant, non_constant
+  s1_const, s1_non_const = split_constants(s1)
+  s2_const, s2_non_const = split_constants(s2)
+  if s1_const == s2_const == 0:
+    return 1
+  for s2_d in s2_non_const:
+    def remove_first(s1_non_const):
```

Comment:
This code could be simplified if we decide we don't support reshape with multiple occurrences of a dimension and if we don't support `reshape` with `-1` dimensions. In that case, the only use for `divide_shape_sizes` really is to compare the shape sizes for equality.